apps/server/src/modules/oauth/service/oauth.service.ts

```diff
@@ -0,0 +1,160 @@
+
+@Injectable()
+export class OAuthService {
+	constructor(
+		private readonly userRepo: UserRepo,
+		private readonly systemRepo: SystemRepo,
+		private readonly jwtService: FeathersJwtProvider,
+		private httpService: HttpService,
+		@Inject('OAuthEncryptionService') private readonly oAuthEncryptionService: SymetricKeyEncryptionService,
+		private iservOauthService: IservOAuthService,
+		private logger: Logger
+	) {
+		this.logger.setContext(OAuthService.name);
+	}
+
+	/**
+	 * @query query input that has either a code or an error
+	 * @return authorization code or throws an error
+	 */
+	checkAuthorizationCode(query: AuthorizationParams): string {
+		if (query.code) return query.code;
+		let errorCode = 'sso_auth_code_step';
+		if (query.error) {
+			this.logger.log(query.error);
+			errorCode = `sso_oauth_${query.error}`;
+			this.logger.error(`SSO Oauth authorization code request return with an error: ${query.code as string}`);
+		}
+		throw new OAuthSSOError('Authorization Query Object has no authorization code or error', errorCode);
+	}
+
+	async requestToken(code: string, system: System): Promise<OauthTokenResponse> {
+		const decryptedClientSecret: string = this.oAuthEncryptionService.decrypt(system.oauthConfig.clientSecret);
+		const tokenRequestPayload: TokenRequestPayload = TokenRequestPayloadMapper.mapToResponse(
+			system,
+			decryptedClientSecret,
+			code
+		);
+		const responseTokenObservable = this.httpService.post<OauthTokenResponse>(
+			tokenRequestPayload.tokenEndpoint,
+			QueryString.stringify(tokenRequestPayload.tokenRequestParams),
+			{
+				method: 'POST',
+				headers: {
+					'Content-Type': 'application/x-www-form-urlencoded',
+				},
+			}
+		);
+		const responseToken = await lastValueFrom(responseTokenObservable);
+		return responseToken.data;
+	}
+
+	async _getPublicKey(system: System): Promise<string> {
+		const client: JwksRsa.JwksClient = JwksRsa({
+			cache: true,
+			jwksUri: system.oauthConfig.jwksEndpoint,
+		});
+		const key: JwksRsa.SigningKey = await client.getSigningKey();
+		return key.getPublicKey();
+	}
+
+	async validateToken(idToken: string, system: System): Promise<IJwt> {
+		const publicKey = await this._getPublicKey(system);
+		const verifiedJWT = jwt.verify(idToken, publicKey, {
+			algorithms: ['RS256'],
+			issuer: system.oauthConfig.issuer,
+			audience: system.oauthConfig.clientId,
+		});
+		if (typeof verifiedJWT === 'string' || verifiedJWT instanceof String)
+			throw new OAuthSSOError('Failed to validate idToken', 'sso_token_verfication_error');
+		return verifiedJWT as IJwt;
+	}
+
+	async findUser(decodedJwt: IJwt, systemId: string): Promise<User> {
+		let user: User;
+		const system: System = await this.systemRepo.findById(systemId);
+		// iserv strategy
+		if (system.oauthConfig.provider === 'iserv') {
+			const uuid = this.iservOauthService.extractUUID(decodedJwt);
+			user = await this.iservOauthService.findUserById(uuid, systemId);
+			return user;
+		}
+		// TODO in general
```

Comment:
when no known strategy can be used, an error should be raised. Avoid any undefined behavior.

Reply:
done.